Document release process

haszari opened the pull request:
The goal of this PR is to capture the intent and outcomes of the phases of releasing the plugin, and include enough detail so a competent team member can confidently captain a release.

The focus of the doc is the high-level steps of release and what they should achieve. Details of how to do each step are included, but these may change. 

Note that this doc is intended to be authoritative – if some of this is covered in handbook or elsewhere, we should update other docs to link to here.

Please read through the doc changes, as if you were following the process for the first time.

1. [Click here to start from CONTRIBUTING.md](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/release-process/CONTRIBUTING.md), then scroll down to Release section & follow link. (Note - deleted sections from this file.)
1. or [Click here to read new RELEASE.md](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/release-process/RELEASE.md).

- Does this clearly explain the expected outcomes of the release process?
- When you do a release, is this what you do? If not, let's tweak so we're all following same process.
- Is there anything missing, in particular any high level/intent info?
- Is this process accurate for major/minor releases? (I've only done one fix release, limited experience!)
- Is there any specific detailed guidance/tips that we need to add (i.e. cheat sheet)?
- Do you refer to any other docs when doing a release? If so, let's ensure this doc covers everything.
- Is this broadly consistent with other release processes at Automattic?

There is a todo section at the bottom with some related info we could document or link to (as follow up - doesn't need to block PR). Some of this may be covered by handbook page (on p2).

I'd like to add a rough timeline to this if possible – so we document the rough "cadence" of releases. Ideas welcome :)

timmyc commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+### Prepare release
+#### Ensure release branch readme is up to date
+- Run changelog script `npm run changelog` to get changelog txt for readme. Changelog content will be output to screen by script.
+- Add changelog section for release, e.g. `= 2.5.11 - 2020-01-20 =`.
```
You could link to an example of this in the readme.txt too.

timmyc commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+_Outcome_: __`readme.txt` is up to date with changes for release.__
+
+#### Build zip & smoke test
+- Ensure you are on tip of release branch.
```
Just to clarify, this is confirming you have `git pull origin release/X.X`? I guess one way to remove uncertainty of this in the doc is to include all relevant git commands and have an example branch name throughout the doc too.

timmyc commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+  - Recommend doing this in a clean environment, e.g. Jurassic.Ninja site.
+  - Sanity check - check correct blocks exist and function.
+  - Test to confirm new features/fixes are working correctly.
+  - Smoke test – test a cross section of core functionality.
```
Is it considered best practice to have another team member smoke test too?

senadir replied:
Majority of release bugs have been discovered by peers, so having someone else testing should be best practice

haszari replied:
This makes sense to me too – I've added a note and "recommended" is bold :) 772229d06cdace88c984a1a3a6188ea34026af0e

As our release process takes shape, I'm getting a better idea of the time/people needed for a release. I think the process doc should help guide us here - let's keep iterating on this.

timmyc commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+- Push release to WPORG `npm run release`.
+- Note that this may happen a day or so after Github release.
+- FYI: Sometimes we make small changes (e.g. readme only) by [committing directly to svn repo](https://developer.wordpress.org/plugins/wordpress-org/how-to-use-subversion/#editing-existing-files).
+
```
Should we add a step here to verify the release looks good on WPORG svn? I know sometimes it takes a little bit for the new release to "appear" as available on a site that has the plugin installed.

nerrad replied:
> Should we add a step here to verify the release looks good on WPORG svn?

Good catch! We definitely want to add that step here.  It's also important to make sure that the `tag` is created on wp.org. We had one release where trunk was updated but not the tag folder. WooCommerce.com links to the tag folder for the tagged zip (not sure why, it should just link to latest).

haszari replied:
Added a whole lot more info to this section, added specific things to validate/test and also clarified the two steps for pushing to wporg (run script, then run generated command).

46cb204cb970a4f1165752fc853109237f2a316e

Having done this section for the first time this morning certainly helped me flesh out the details here!!

timmyc commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+- Relationship to Woo core release
+- Testing blocks plugin as included in Woo core (do we do this?)
+- Policy for inclusion in major / minor / fix releases - i.e. how we use versioning 
+- ["Legacy"](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/release-process/assets/js/legacy/README.md) stuff - is this relevant to releases?
```
Another item that came up recently is around ensuring all Woo core packages ( Blocks, WC-Admin, REST API ) are running the same version of the jetpack-autoloader. So that might be another good one to add to the todo list here... thought we have yet to determine the workflow around when that should be updated across all 3 repos.

nerrad replied:
> Another item that came up recently is around ensuring all Woo core packages ( Blocks, WC-Admin, REST API ) are running the same version of the jetpack-autoloader.

Great to bring this up, thanks! However, this will be tricky to consider as a part of the release process because you could get a specific release in alignment version wise but then either of the other repositories could bump jetpack autoloader version before the release is pulled in at some later time (at which point the problem will surface anyways).  This is another reason why I think we have to switch to ranges for jetpack autoloader version instead of pinning to a specific version.  I think we should leave that out of this iteration of the release notes until we've resolved how this will work (and hopefully won't be something needing thought about in a release).

haszari replied:
🐲 

I've added mention of this to that section. Note that this is not a list of things that the release process should document - it's a list of things that might be related and might need clarification/documenting somewhere.

nerrad replied:
> I've added mention of this to that section. Note that this is not a list of things that the release process should document - it's a list of things that might be related and might need clarification/documenting somewhere.

I appreciate having this list but I don't see the need to keep it in the release documentation since it's not relevant to someone following instructions for doing a release.  Some of these items are also listed as questions which can be answered without documenting.  If you see any of these as important things that we should handle, then create an issue for it.  However, to help with deciding whether something needs an issue (or faq documentation somewhere), I've gone through and addressed the items below 👇 

**Release milestones**

I don't know what you mean by this item.  Do you mean documenting when release milestones get created? Do you mean how release milestones differ from Zenhub releases?

**Relationship to Woo core release**

This is another bullet point that seems vague for a todo.  I think we already have some info on this in the project's README.md: 

> Feature plugin for WooCommerce + Gutenberg. This plugin serves as a space to iterate and explore new Blocks for WooCommerce, and how WooCommerce might work with the block editor.
>
> If you're just looking to use these blocks, update to the latest version of WooCommerce! The blocks are bundled into WooCommerce since version 3.6, and can be added from the "WooCommerce" section in the block inserter.

**Testing blocks plugin as included in Woo core (do we do this?)**

If you mean our squad, we do and we don't.  We do, because what is in woo core is exactly what is in the tagged release (because the tag is pulled in via composer).  So essentially, _yes, it's been tested_. We don't, because we don't explicitly test the pull request that bumps the woo blocks version in composer.json for Woo Core.  I don't think it's necessary for our squad to do because when our plugin is bumped in Woo Core, there is testing done before Woo core is published.  Any issues with blocks will either be bugs we didn't catch in our own testing (i.e. not specifically from including in woo core) or because something went wrong with the woo core release process not pulling in the woo blocks package correctly (which is not something specific to the plugin itself).

**Policy for inclusion in major / minor / fix releases - i.e. how we use versioning**

I don't think we should be too rigid here when it comes to policy, but it looks like we pretty much follow semver for the most part.

- patch: important fixes that we earmark for including in a minor version release of WC core.  These might also go on wp.org if the minor version on wp.org is the same branch as the patch.
- minor: new features and bug fixes released for the plugin are typically minors. Minor's will get bundelled in whatever version of WC core they are earmarked for.
- major: breaking changes. These would be included in a version of WC core that is a major release (breaking release).

**["Legacy"](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/release-process/assets/js/legacy/README.md) stuff - is this relevant to releases?**

You never addressed [my comment here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1614/files#r368958263) which I think is relevant for testing, but when it comes to the question "is this relevant to releases?" the answer is other than testing, no.  I also don't want to invest too heavily in documenting this (besides the [documentation we already have](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/master/assets/js/legacy)) until we've had a chance to see this system in use for a while.

**`jetpack-autoloader` versions in this plugin vs. wc-admin vs. WooCommerce core**

I don't think this should be a todo in a "how to release document" for the reasons I've already given. Instead I think it should just be an issue to deal with (see #1625).

haszari replied:
Thanks for clarifying those points @nerrad - lots of useful info in this comment. I agree this is not the right doc for this content; it all came up for me as things that might be useful in our team docs somewhere. Removing this section now :)

As for what next with this info – let's follow that up in team channel. I'm hesitant to add docs issues for every question I have :)

Aljullu commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+##### Release to WPORG
+- Push release to WPORG `npm run release`.
+- Note that this may happen a day or so after Github release.
+- FYI: Sometimes we make small changes (e.g. readme only) by [committing directly to svn repo](https://developer.wordpress.org/plugins/wordpress-org/how-to-use-subversion/#editing-existing-files).
```
It's quite similar to the page you linked, but in `CONTRIBUTING.md` we have these steps detailed as well:
https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/CONTRIBUTING.md#updating-a-single-file-in-wordpressorg
The nice thing is that commands are already with our repo URL, so it's just copy & pasting into the terminal.

Do you think it might be worth keeping it or you removed it in purpose?

nerrad replied:
> Do you think it might be worth keeping it or you removed it in purpose?

I agree with Albert here, it'd be nice to keep the example commands for updating the readme.txt here.

haszari replied:
> Do you think it might be worth keeping it or you removed it in purpose?

Yes, I did remove this on purpose :) To me it didn't seem essential for us to document this – it's just committing a change to a file (we're not special). The key thing was to note that this sometimes happens and is part of how we do things. Also it's a little distracting having this in the release process.

So .. I've reinstated the previous doc (I don't think it hurts to include it) as an "appendix" on the release page. 3ba987f0a513575ab2a7fec82bef145602201ead

nerrad commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+    - `read:user`
+  - Ensure your token is available to scripts, e.g. `export GH_API_TOKEN={YOUR-TOKEN}` or similar.
+- Get WPORG plugin svn commit access - ask a team member.
+
```
>  I realize the use of npm is generally a given but perhaps it could be worth considering mentioning (somewhere - maybe not here) that it is needed as a baseline for development of the plugin and what versions are recommended etc.

This is a very good point. Along with mentioning the requirement for npm/node, we should point to the `engines` property in our `package.json` to point out the required minimum versions.

haszari replied:
Ha! Missing dev docs! I've added this to CONTRIBUTING. 

I also added a hint/link to package.json, but that feels overly detailed. Ideally we'd have some high-level guidance here, e.g. "we recommend using node LTS or better".

nerrad commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+## Release process
+### Lead-up to release
+#### Ensure release development is complete
+- Release milestone should exist (!).
```
> This only applies to point releases, right? For minor and major we are using ZenHub.

I noticed the same thing here. I think there's a few other places where Zenhub can be mentioned as well. Currently, Zenhub releases contain all the pulls for major and minor releases and milestones are only used for _patch_ releases.

haszari replied:
Thanks for the info - I don't have any experience of WC blocks major/minor releases :)

I've added a section at the bottom defining version terms and corresponding "types" of release. (This is possibly the start of a different doc on how we use versioning, release schedule/strategy etc.)

Where the steps differ, I've consistently signalled this, and settled on the terms `major` `minor` and `patch`.

I've left one specific TBC about zenhub, i.e. a placeholder for that side of the process. @nerrad  or anyone feel free to commit some tweaks to these steps to clarify things. We can iterate on this more after merge of course :) b8b0e1284ef4cefb37159ee2c43c19dbbb703fea

nerrad commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+  - Ideally, use GitHub Hub to cherry pick the PR - `hub am -3 {http://URL-TO-PR}`. 
+  - If there are serious conflicts or extensive differences between `master` and release branch, you may need to take more care:
+    - Manually cherry pick individual commits using git - `git cherry-pick {COMMIT-HASH}`.
+    - Or in some cases, manually craft a new PR with appropriate changes, targeting release branch.
```
I love how you've broken down the potential ways a pr might be "cherry-picked" into the release branch. It's a good idea to highlight there may be times where it's not a clean merge.

senadir replied:
`hub am -3` is equivalence to `git am -3` which is equivalence to a graceful cherry-pick, you would never need to cherry-pick manually, since `am -3` will prompt you to fix conflict anytime there is one.

haszari replied:
Leaving this as is – with cherry picking and furious parallel development 🎆  on `master` there's always a risk that we'll need to take care with changes, and sometimes manually apply them. Cherry picking individual commits may still be a useful approach even if it's technically similar to cherry picking a multiple commits in a PR with `hub am -3`.

As we get more experience using this process we can finesse the wording of these steps so they are super clear and helpful.

nerrad commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+
+### Prepare release
+#### Ensure release branch readme is up to date
+- Run changelog script `npm run changelog` to get changelog txt for readme. Changelog content will be output to screen by script.
```
This will need updated once @senadir or myself improve the changelog to allow for correctly selecting either Zenhub releases or Github milestones as the source for changelog entries. More than likely we'll have this configured to use cli flags and update the scripts in `package.json` so we'd have `npm run changelog:zh` and `npm rum changelog:gh` variations. I'm just noting this for now, we could publish as is and just update when we get the changelog update done.

haszari replied:
Thanks for your vigilance! We can update this later. This is a living doc, this is just the first iteration.

Note also from my point of view, the "cheatsheet" info (actual commands) is less important to document, compared to the phases and intentions/outcomes of the release steps. 

As we road-test the document on more releases, and/or as our tooling evolves, some of these details may be hard to maintain and get removed – I think this is ok. If the outcomes and phases are relatively stable then the doc is doing the job IMO 🦄.

nerrad commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+- Update dependencies – `npm ci`.
+- Run a production build - `npm run build`.
+- Run package script to get a zip to test `npm run package-plugin`.
+- Smoke test built release zip:
```
I think it's also important to test on an existing site with blocks that were created before this release. This ensures that there are no unexpected block invalidation errors when editing that pre-existing content.

haszari replied:
Yes! 💌 this. 

As I get more details on all the testing involved in a release, it's starting to look like a significant chunk of time. Maybe as a next step we can map out how this looks in terms of time spent and people involved.

haszari replied:
Love this – I've now placed the testing emphasis on testing the zip before github/wporg release. It makes sense that we do the most thorough testing early, and the testing after that is quicker – confidence check to confirm no build / upload problems, test cross section of functionality.

nerrad commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+
+_Outcomes_: __Version numbers updated in source code & developers can test tagged release.__
+
+##### Release to WPORG
```
This probably should be at the same heading level as `Release to Github`?

nerrad commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+
+##### Release to WPORG
+- Push release to WPORG `npm run release`.
+- Note that this may happen a day or so after Github release.
```
Not sure this is needed.

senadir replied:
> - Note that this may happen a day or so after Github release.
yeah, usually we release within 1 hour after the GitHub release, just to give us time to test the GH tag again before pushing to users.

haszari replied:
I've removed this for now.

The key priority in this PR is getting the overall process documented. As follow up we can look at timeframe for each phase, and consider documenting that somewhere.

nerrad commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+- Relationship to Woo core release
+- Testing blocks plugin as included in Woo core (do we do this?)
+- Policy for inclusion in major / minor / fix releases - i.e. how we use versioning 
+- ["Legacy"](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/release-process/assets/js/legacy/README.md) stuff - is this relevant to releases?
```
The legacy stuff is handled by the build process so there's nothing specific the person doing the release has to do. However this does raise the point that as long as we still officially support WP 5.0, we should be testing against that version of WordPress to make sure the blocks work as expected (I think we often forget that step). Currently on versions of WP < 5.3, All Products, and the filter blocks won't be available there, but all our other "legacy" blocks should still work fine.

haszari replied:
Thanks for reminding me about this – I didn't quite understand yesterday :) 

I've added this to the pre-release final testing of the zip bbba14ded3a23d12bdf8437c0c9dd6c8cff88eb5:

`Test to confirm blocks are available and work correctly in oldest supported WordPress version (e.g. 5.0).`

Aljullu commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+
+_Outcome_: __Customers can install/update via WPORG; WPORG plugin page is up to date__.
+
+### After release
```
One thing we should probably add to `After release` if it's a point release is to close the milestone from:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/milestones

haszari commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+# Releases
+
```
Good idea. Added a really high-level sentence, we can keep iterating on this :)

haszari commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+- Make release branch if needed (major/minor).
+  - For fix releases, the branch should already exist.
+  - Otherwise create branch: `release/X.X`.
+- Update your local checkout to tip of the release branch.
```
Thanks – I was probably aiming for brevity / efficiency over correctness :)

I've made the changes you suggested, we can can iterate on this as we go :)

haszari commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+  - Otherwise create branch: `release/X.X`.
+- Update your local checkout to tip of the release branch.
+- Cherry pick all PRs into the release branch:
+  - If branch is already labelled `status: cherry-picked 🍒` then continue to next PR.
```
Good spotting! Fixed here :) 75b45eae61acd1bc51b497df7060c2de71f1fb33

haszari commented on RELEASE.md:
```diff
@@ -0,0 +1,101 @@
+- Make release branch if needed (major/minor).
+  - For fix releases, the branch should already exist.
+  - Otherwise create branch: `release/X.X`.
+- Update your local checkout to tip of the release branch.
```
Have clarified that cherry picking is for patch releases here: 75b45eae61acd1bc51b497df7060c2de71f1fb33

Re updating branch, this is probably still an important step for other releases? If not I don't think it hurts to include it. The more we can "harmonize" the process the better :)

Aljullu replied:
Yeah... I just found it confusing that step 1 is 'create release branch' and step 2 is 'make sure you are on the tip of release branch' (_of course I am, I just created it!_ :smile: ). But yes, it doesn't hurt so it's fine for me to leave it there.

haszari commented:
Howdy y'all! Great to have so much feedback on this, thank you 👏 

I've added and adjusted things to address all the feedback. Please feel free to take another read through, and feel free to commit changes directly for example you have ideas for wording or can clarify missing details.

Note that I've added new content to [CONTRIBUTING.md](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/release-process/CONTRIBUTING.md) as well as [RELEASE.md](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/release-process/RELEASE.md).

This doc is growing quickly - up to 1000 words now! One of the key goals for me is that this is easy to maintain, and easy to follow when running a release. As we road test the doc and iterate (and our tools evolve?), we may slim down some of the technical details to keep it maintainable and easy to read/navigate.

Aljullu commented on RELEASE.md:
```diff
@@ -0,0 +1,160 @@
+- Add changelog section for release, e.g. [`= 2.5.11 - 2020-01-20 =`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/74a41881bfa456a2167a52aaeb4871352255e328).
+- Copy-paste the changelog content into `readme.txt`.
+- Make any other changes to readme as needed - e.g. support versions changing, new blocks. 
+- Push readme changes to release branch on origin repo.
```
Suggestion to add something at the end of this line: `If you are confident of your changes, no need to create a PR and wait for approval. You can directly push.`. We had some confusion in the past whether this had to be done via a PR or a direct push.

Aljullu commented on RELEASE.md:
```diff
@@ -0,0 +1,160 @@
+#### Update `master` with release changes
+- Ensure changelog is up to date on master.
+- For _major_ & _minor_ releases, update version on master with dev suffix, e.g. [`2.6-dev`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/e27f053e7be0bf7c1d376f5bdb9d9999190ce158).
+- Code changes should already be present on master - all dev happens on `master`.
```
I'm not sure to understand this line. Which code changes is it referring to?

nerrad replied:
I think this refers to ongoing development since we merge all changes to master first.  This line can probably be removed.

Aljullu commented on RELEASE.md:
````diff
@@ -0,0 +1,160 @@
+4. Commit the changes to the server:
+
+```
+svn ci -m "Updated readme.txt description"
````
Missing ```.
